hazelcast-jclouds/src/main/java/com/hazelcast/jclouds/ComputeServiceBuilder.java

```diff
@@ -0,0 +1,282 @@
+import com.hazelcast.config.properties.PropertyDefinition;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.util.Preconditions;
+import org.jclouds.Constants;
+import org.jclouds.ContextBuilder;
+import org.jclouds.compute.ComputeService;
+import org.jclouds.compute.ComputeServiceContext;
+import org.jclouds.compute.domain.ComputeMetadata;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.domain.Credentials;
+import org.jclouds.domain.Location;
+import org.jclouds.googlecloud.GoogleCredentialsFromJson;
+import org.jclouds.location.reference.LocationConstants;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Properties;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+/**
+ * ComputeServiceBuilder is the responsible class for building jclouds compute service provider.
+ * Also parses config and applies necessary filters on cluster nodes.
+ */
+public class ComputeServiceBuilder {
+
+    private static final String JCLOUD_CONNECTION_TIMEOUT = "10000";
+    private static final ILogger LOGGER = Logger.getLogger(ComputeServiceBuilder.class);
+
+    private final Map<String, Comparable> properties;
+    private Set<String> regionsSet = new HashSet<String>();
+    private Set<String> zonesSet = new HashSet<String>();
+    private Set<String> tagsKeysSet = new HashSet<String>();
+    private Set<String> tagsValuesSet = new HashSet<String>();
+    private Predicate<ComputeMetadata> nodesFilter;
+    private ComputeService computeService;
+
+    /**
+     * Instantiates a new Compute service builder.
+     *
+     * @param properties the properties
+     */
+    ComputeServiceBuilder(Map<String, Comparable> properties) {
+        if (properties == null) {
+            throw new NullPointerException("Properties can't be null");
+        }
+        this.properties = properties;
+    }
+
+    /**
+     * Gets filtered nodes.
+     *
+     * @return the filtered nodes
+     */
+    public  Iterable<? extends NodeMetadata> getFilteredNodes() {
+        final String group = getOrNull(JCloudsProperties.GROUP);
+        Set<? extends NodeMetadata> result = computeService.listNodesDetailsMatching(nodesFilter);
+        Iterable<? extends NodeMetadata> filteredResult = new HashSet<NodeMetadata>();
+        for (NodeMetadata metadata : result) {
+            if (group != null && !group.equals(metadata.getGroup())) {
+                continue;
+            }
+            if (checkRegionZoneConfig(metadata)) {
+                continue;
+            }
+            ((HashSet<NodeMetadata>) filteredResult).add(metadata);
+        }
+        return filteredResult;
+    }
+
+    public int getServicePort() {
+        return getOrDefault(JCloudsProperties.HZ_PORT, NetworkConfig.DEFAULT_PORT);
+    }
+
+    private boolean checkRegionZoneConfig(NodeMetadata metadata) {
```

Comment:
i think this method can be implemented in simpler manner without that many branching and local variables.